nectar/downloaders/threaded.py

```diff
@@ -273,119 +273,128 @@ def _fetch(self, request, session):
-            report.download_connection_error()
-
-        except requests.Timeout:
-            """
-            Handle a timeout differently than a connection error. Do not add
-            to failed_netlocs so that a new connection can be attempted.
-            """
-            _logger.warning("Request Timeout - Connection with {url} timed out.".format(
-                url=request.url)
-            )
-            report.download_connection_error()
-
-        except DownloadCancelled as e:
-            _logger.info(str(e))
-            report.download_canceled()
-
-        except DownloadFailed as e:
-            _logger.info('Download failed: %s' % str(e))
-            report.error_msg = e.args[2]
-            report.error_report['response_code'] = e.args[1]
-            report.error_report['response_msg'] = e.args[2]
-            report.download_failed()
-
-        except Exception as e:
-            _logger.exception(e)
-            report.error_msg = str(e)
-            report.download_failed()
+                if ignore_encoding or self.config.stream:
+                    chunks = self.chunk_generator(response.raw, self.buffer_size)
+                else:
+                    chunks = response.iter_content(self.buffer_size)
+
+                for chunk in chunks:
+                    if self.is_canceled or request.canceled:
+                        raise DownloadCancelled(request.url)
+
+                    file_handle.write(chunk)
+
+                    bytes_read = len(chunk)
+                    report.bytes_downloaded += bytes_read
+
+                    now = datetime.datetime.now()
+
+                    if now - last_update_time >= progress_interval:
+                        last_update_time = now
+                        self.fire_download_progress(report)
+
+                    with self._bytes_lock:
+                        if now - self._time_bytes_this_second_was_cleared >= ONE_SECOND:
+                            self._bytes_this_second = 0
+                            self._time_bytes_this_second_was_cleared = now
+                        self._bytes_this_second += bytes_read
+
+                    if max_speed is not None and self._bytes_this_second >= max_speed:
+                        # it's not worth doing fancier mathematics than this, very
+                        # fine-grained sleep times [1] are not honored by the system
+                        # [1] for example, sleeping the remaining fraction of time
+                        # before this second is up
+                        time.sleep(0.5)
+
+                # guarantee 1 report at the end
+                self.fire_download_progress(report)
+
+            except SkipLocation:
+                _logger.debug("Skipping {url} because {netloc} could not be reached.".format(
+                    url=request.url, netloc=netloc)
+                )
+                report.download_skipped()
+
+            except requests.ConnectionError as e:
```

Comment:
should not we just leave the re-try logic for ConnectionError only?

Reply:
I think connection reset from server side actually doesn't produce connection error. Connection error should be retried by urlllib

Reply:
The connection reset that we want to solve is server issue a RST when data is streaming. This is not covered by the retry in urllib so I think we should only retry when the "response.status_code" is not set.

My suggestion is to check the "response.status_code" in the "DownloadFailed" exception (https://github.com/pulp/nectar/pull/72/files#diff-609b5250f930300d62fd3da027ef9608R368). Retry only if status_code is not set.

You can refer my patch in the pulp_docker:
https://github.com/pulp/pulp_docker/pull/459/files#diff-cebb975e7cc2d51eddda339814fb7bd2R477